app/vmalert/rule/alerting.go

```diff
@@ -733,3 +637,33 @@ func (ar *AlertingRule) alertsToSend(ts time.Time, resolveDuration, resendDelay
 	}
 	return alerts
 }
+
+// GetAlerts returns active alerts of rule
+func (ar *AlertingRule) GetAlerts() []*notifier.Alert {
+	ar.alertsMu.RLock()
+	defer ar.alertsMu.RUnlock()
+	var alerts []*notifier.Alert
+	for _, a := range ar.alerts {
+		alerts = append(alerts, a)
+	}
+	return alerts
+}
+
+// GetAlert returns alert if id exists
+func (ar *AlertingRule) GetAlert(id uint64) *notifier.Alert {
+	ar.alertsMu.RLock()
+	defer ar.alertsMu.RUnlock()
+	if ar.alerts == nil {
+		return nil
+	}
+	return ar.alerts[id]
+}
+
+// UpdateRuleAlerts update alert with given hash id
+// Isn't concurrent-safe, callers must be sure to call it from only one goroutine
+func (ar *AlertingRule) UpdateRuleAlerts(id uint64, alert *notifier.Alert) {
```

Comment:
this shouldn't be public. I'd recommend removing this method.
If it is needed only for web_test, I'd suggest rewriting the tests instead. It should be feasible to create a group via public method with fake querier and call execution once.

Reply:
Right, my bad!
Fixed, plz check.